token/services/nftcc/uniqueness/uniqueness.go

```diff
@@ -0,0 +1,84 @@
+/*
+Copyright IBM Corp. All Rights Reserved.
+
+SPDX-License-Identifier: Apache-2.0
+*/
+
+package uniqueness
+
+import (
+	"crypto/rand"
+	"crypto/sha256"
+	"encoding/base64"
+	"encoding/json"
+	"github.com/hyperledger-labs/fabric-smart-client/platform/view"
+	"github.com/hyperledger-labs/fabric-smart-client/platform/view/services/kvs"
+	"github.com/pkg/errors"
+	"sync"
+)
+
+// Service is a uniqueness service.
+// The service computes a unique id for a given object by hashing the object's json representation together with a random salt.
+// The random salt is used to avoid dictionary attacks.
+// The random salt is stored in the KVS and is generated on the first call to the service.
+type Service struct {
+	mutex sync.Mutex
+	sp    view.ServiceProvider
+}
+
+// ComputeID computes the unique ID of the given object.
+func (s *Service) ComputeID(state interface{}) (string, error) {
+	if state == nil {
+		return "", errors.New("state is nil")
+	}
+
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+
+	kvs := kvs.GetService(s.sp)
+	k := "github.com/hyperledger-labs/fabric-token-sdk/token/services/nftcc/uniqueness/key"
+	var key []byte
+	if kvs.Exists(k) {
+		if err := kvs.Get(k, &key); err != nil {
+			return "", errors.WithMessagef(err, "failed to get key %s", k)
+		}
+	} else {
+		// sample a random 32 bytes key and store it
+		size := 32
+		key = make([]byte, size)
+		n, err := rand.Read(key)
+		if err != nil {
+			return "", errors.Wrap(err, "error getting random bytes")
+		}
+		if n != size {
+			return "", errors.New("error getting random bytes")
+		}
+		if err := kvs.Put(k, key); err != nil {
+			return "", errors.WithMessagef(err, "failed to put key %s", k)
+		}
```

Comment:
What about the use of `errors.Wrap` and `errors.WithMessagef`. For all those errors here, does it make sense to use `errors.Wrap` to also extend the stack trace, rather than only extending the error message?

Reply:
I'm using WithMessagef because I expect the kvs to already have the stack attached. So, I just want to append the message. But, I will adjust the other lines that are not in sync